File: backend/src/backend/api/routes/schematics.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Header, HTTPException, Response, status

from backend.database import get_schematic_repository
from backend.repositories.schematics import SchematicRepository
from backend.schemas.schematic import SchematicCreate, SchematicResponse, SchematicSummary, SchematicUpdate

router = APIRouter(prefix="/api/v1/schematics", tags=["schematics"])


def require_user_id(x_user_id: Annotated[str | None, Header()] = None) -> str:
    # Temporary Feature 1 ownership boundary. Feature 4 auth should replace this
    # header with the authenticated user id.
    if not x_user_id or not x_user_id.strip():
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail={
                "error_code": "E400",
                "message": "X-User-Id header is required until authentication is implemented.",
            },
        )
    return x_user_id.strip()


@router.get("", response_model=list[SchematicSummary])
def list_schematics(
    user_id: Annotated[str, Depends(require_user_id)],
    repository: Annotated[SchematicRepository, Depends(get_schematic_repository)],
) -> list[SchematicSummary]:
    return [
        SchematicSummary(id=document["id"], name=document["name"], updated_at=document["updated_at"])
        for document in repository.list(user_id)
    ]


@router.post("", response_model=SchematicResponse, status_code=status.HTTP_201_CREATED)
def create_schematic(
    payload: SchematicCreate,
    user_id: Annotated[str, Depends(require_user_id)],
    repository: Annotated[SchematicRepository, Depends(get_schematic_repository)],
) -> SchematicResponse:
    return SchematicResponse(**repository.create(user_id, payload))


@router.get("/{schematic_id}", response_model=SchematicResponse)
def get_schematic(
    schematic_id: str,
    user_id: Annotated[str, Depends(require_user_id)],
    repository: Annotated[SchematicRepository, Depends(get_schematic_repository)],
) -> SchematicResponse:
    document = repository.get(user_id, schematic_id)
    if document is None:
        raise_not_found(schematic_id)
    return SchematicResponse(**document)


@router.put("/{schematic_id}", response_model=SchematicResponse)
def update_schematic(
    schematic_id: str,
    payload: SchematicUpdate,
    user_id: Annotated[str, Depends(require_user_id)],
    repository: Annotated[SchematicRepository, Depends(get_schematic_repository)],
) -> SchematicResponse:
    document = repository.update(user_id, schematic_id, payload)
    if document is None:
        raise_not_found(schematic_id)
    return SchematicResponse(**document)


@router.delete("/{schematic_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_schematic(
    schematic_id: str,
    user_id: Annotated[str, Depends(require_user_id)],
    repository: Annotated[SchematicRepository, Depends(get_schematic_repository)],
) -> Response:
    deleted = repository.delete(user_id, schematic_id)
    if not deleted:
        raise_not_found(schematic_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


def raise_not_found(schematic_id: str) -> None:
    raise HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail={
            "error_code": "E404",
            "message": f"Schematic '{schematic_id}' does not exist or you do not have permission to access it.",
        },
    )

File: mern-backend/src/routes/schematics.ts
import { Router } from "express";

import { SchematicRepository } from "../repositories/schematics.js";
import { normalizeSchematicPayload, ValidationError } from "../schemas/schematic.js";
import { ApiError } from "../utils/errors.js";

export const schematicsRouter = Router();
const repository = new SchematicRepository();

const requireUserId = (value: unknown) => {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new ApiError(401, "E400", "X-User-Id header is required until authentication is implemented.");
  }
  return value.trim();
};

const raiseNotFound = (schematicId: string) => {
  throw new ApiError(
    404,
    "E404",
    `Schematic '${schematicId}' does not exist or you do not have permission to access it.`,
  );
};

schematicsRouter.get("/", async (request, response, next) => {
  try {
    const userId = requireUserId(request.header("X-User-Id"));
    response.json(await repository.list(userId));
  } catch (error) {
    next(error);
  }
});

schematicsRouter.post("/", async (request, response, next) => {
  try {
    const userId = requireUserId(request.header("X-User-Id"));
    const payload = normalizeSchematicPayload(request.body);
    response.status(201).json(await repository.create(userId, payload));
  } catch (error) {
    if (error instanceof ValidationError) next(new ApiError(422, "E200", error.message));
    else next(error);
  }
});

schematicsRouter.get("/:schematicId", async (request, response, next) => {
  try {
    const userId = requireUserId(request.header("X-User-Id"));
    const document = await repository.get(userId, request.params.schematicId);
    if (!document) raiseNotFound(request.params.schematicId);
    response.json(document);
  } catch (error) {
    next(error);
  }
});

schematicsRouter.put("/:schematicId", async (request, response, next) => {
  try {
    const userId = requireUserId(request.header("X-User-Id"));
    const payload = normalizeSchematicPayload(request.body);
    const document = await repository.update(userId, request.params.schematicId, payload);
    if (!document) raiseNotFound(request.params.schematicId);
    response.json(document);
  } catch (error) {
    if (error instanceof ValidationError) next(new ApiError(422, "E200", error.message));
    else next(error);
  }
});

schematicsRouter.delete("/:schematicId", async (request, response, next) => {
  try {
    const userId = requireUserId(request.header("X-User-Id"));
    const deleted = await repository.delete(userId, request.params.schematicId);
    if (!deleted) raiseNotFound(request.params.schematicId);
    response.status(204).send();
  } catch (error) {
    next(error);
  }
});
